### ISSUE TITLE
Add hyperparameter configuration

### ISSUE THREAD

**Issue by tgrunnagle** [tgrunnagle/go-towerfall#28]
## Overview

**Task ID:** TASK-024
**Parent Task:** TASK-P006 (PPO Agent Implementation)
**Epic:** #14

This task creates a configuration system for PPO hyperparameters to enable flexible training runs with different parameter settings. A well-designed configuration system allows for easy experimentation, reproducibility, and tuning of the PPO agent.

## Context

The PPO (Proximal Policy Optimization) algorithm has numerous hyperparameters that significantly impact training performance. Rather than hardcoding these values, we need a robust configuration system that:
- Provides sensible defaults based on PPO best practices
- Allows easy override via configuration files (YAML)
- Supports validation of parameter ranges
- Enables reproducible training runs

## Dependencies

| Task ID | Issue | Title | Status |
|---------|-------|-------|--------|
| TASK-007 | #25 | Configure bot2 project dependencies | Required |

This task depends on #25 for Pydantic (used for config validation) and the uv environment setup.

### Core PPO Parameters

| Parameter | Type | Default | Description |
|-----------|------|---------|-------------|
| `learning_rate` | float | 3e-4 | Learning rate for Adam optimizer |
| `clip_range` | float | 0.2 | PPO clipping parameter (epsilon) |
| `clip_range_vf` | float or None | None | Value function clipping (None = no clipping) |
| `n_epochs` | int | 10 | Number of epochs per PPO update |
| `batch_size` | int | 64 | Minibatch size for training |
| `n_steps` | int | 2048 | Number of steps to collect per rollout |
| `gamma` | float | 0.99 | Discount factor |
| `gae_lambda` | float | 0.95 | GAE lambda for advantage estimation |
| `ent_coef` | float | 0.01 | Entropy coefficient for exploration |
| `vf_coef` | float | 0.5 | Value function loss coefficient |
| `max_grad_norm` | float | 0.5 | Maximum gradient norm for clipping |

### Network Architecture Parameters

| Parameter | Type | Default | Description |
|-----------|------|---------|-------------|
| `hidden_sizes` | list[int] | [64, 64] | Hidden layer sizes for policy/value networks |
| `activation` | str | "tanh" | Activation function (tanh, relu, elu) |
| `share_features` | bool | True | Whether actor and critic share feature layers |
| `ortho_init` | bool | True | Use orthogonal initialization |

### Training Parameters

| Parameter | Type | Default | Description |
|-----------|------|---------|-------------|
| `total_timesteps` | int | 1_000_000 | Total training timesteps |
| `seed` | int or None | None | Random seed for reproducibility |
| `device` | str | "auto" | Device (cpu, cuda, auto) |
| `normalize_advantage` | bool | True | Normalize advantages per minibatch |
| `target_kl` | float or None | None | Target KL divergence for early stopping |

### Logging and Checkpointing

| Parameter | Type | Default | Description |
|-----------|------|---------|-------------|
| `log_interval` | int | 10 | Frequency of logging (in updates) |
| `save_interval` | int | 50 | Frequency of model checkpoints (in updates) |
| `tensorboard` | bool | True | Enable TensorBoard logging |
| `log_dir` | str | "logs/" | Directory for logs and checkpoints |

### 1. Pydantic Config Classes

Create type-safe configuration models using Pydantic v2:

```python
# bot2/config/ppo_config.py
from pydantic import BaseModel, Field, field_validator
from typing import Literal

class PPOCoreConfig(BaseModel):
    """Core PPO algorithm hyperparameters."""
    learning_rate: float = Field(3e-4, gt=0, description="Learning rate for optimizer")
    clip_range: float = Field(0.2, gt=0, le=1.0, description="PPO clipping parameter")
    clip_range_vf: float | None = Field(None, ge=0, description="Value function clipping")
    n_epochs: int = Field(10, ge=1, description="Epochs per PPO update")
    batch_size: int = Field(64, ge=1, description="Minibatch size")
    n_steps: int = Field(2048, ge=1, description="Steps per rollout")
    gamma: float = Field(0.99, ge=0, le=1.0, description="Discount factor")
    gae_lambda: float = Field(0.95, ge=0, le=1.0, description="GAE lambda")
    ent_coef: float = Field(0.01, ge=0, description="Entropy coefficient")
    vf_coef: float = Field(0.5, ge=0, description="Value function coefficient")
    max_grad_norm: float = Field(0.5, ge=0, description="Max gradient norm")

class NetworkConfig(BaseModel):
    """Neural network architecture configuration."""
    hidden_sizes: list[int] = Field([64, 64], min_length=1)
    activation: Literal["tanh", "relu", "elu"] = "tanh"
    share_features: bool = True
    ortho_init: bool = True

class TrainingConfig(BaseModel):
    """Training loop configuration."""
    total_timesteps: int = Field(1_000_000, ge=1)
    seed: int | None = None
    device: Literal["cpu", "cuda", "auto"] = "auto"
    normalize_advantage: bool = True
    target_kl: float | None = Field(None, gt=0)

class LoggingConfig(BaseModel):
    """Logging and checkpointing configuration."""
    log_interval: int = Field(10, ge=1)
    save_interval: int = Field(50, ge=1)
    tensorboard: bool = True
    log_dir: str = "logs/"

class PPOConfig(BaseModel):
    """Complete PPO configuration."""
    core: PPOCoreConfig = Field(default_factory=PPOCoreConfig)
    network: NetworkConfig = Field(default_factory=NetworkConfig)
    training: TrainingConfig = Field(default_factory=TrainingConfig)
    logging: LoggingConfig = Field(default_factory=LoggingConfig)

    @classmethod
    def from_yaml(cls, path: str) -> "PPOConfig":
        """Load configuration from YAML file."""
        import yaml
        with open(path) as f:
            data = yaml.safe_load(f)
        return cls.model_validate(data)

    def to_yaml(self, path: str) -> None:
        """Save configuration to YAML file."""
        import yaml
        with open(path, "w") as f:
            yaml.dump(self.model_dump(), f, default_flow_style=False)
```

### 2. YAML Configuration File Format

Support human-readable YAML configuration:

```yaml
# config/ppo_default.yaml
core:
  learning_rate: 0.0003
  clip_range: 0.2
  n_epochs: 10
  batch_size: 64
  n_steps: 2048
  gamma: 0.99
  gae_lambda: 0.95
  ent_coef: 0.01
  vf_coef: 0.5
  max_grad_norm: 0.5

network:
  hidden_sizes: [64, 64]
  activation: tanh
  share_features: true
  ortho_init: true

training:
  total_timesteps: 1000000
  device: auto
  normalize_advantage: true

logging:
  log_interval: 10
  save_interval: 50
  tensorboard: true
  log_dir: logs/
```

### 3. Add PyYAML Dependency

Add `pyyaml>=6.0` to `bot2/pyproject.toml` dependencies for YAML support.

### 4. File Structure

```
bot2/
  config/
    __init__.py
    ppo_config.py      # Pydantic config models
    defaults/
      ppo_default.yaml
      ppo_fast.yaml    # Config for fast experimentation
```

## Acceptance Criteria

- [ ] `PPOConfig` Pydantic model defined with all hyperparameters listed above
- [ ] Nested config structure for organization (core, network, training, logging)
- [ ] Field validation with appropriate constraints (gt, ge, le, literal types)
- [ ] `from_yaml()` class method to load configuration from YAML files
- [ ] `to_yaml()` method to save configuration for reproducibility
- [ ] Default configuration file `config/defaults/ppo_default.yaml` provided
- [ ] PyYAML added to project dependencies
- [ ] Unit tests for config validation (valid configs, invalid values, edge cases)
- [ ] Type hints using native Python typing (not `typing.Optional`, etc.)
- [ ] Config can be instantiated with all defaults: `PPOConfig()`

## Testing Requirements

```python
# test_ppo_config.py
def test_default_config():
    """Test that default config is valid."""
    config = PPOConfig()
    assert config.core.learning_rate == 3e-4
    assert config.core.clip_range == 0.2

def test_config_validation():
    """Test that invalid values are rejected."""
    with pytest.raises(ValidationError):
        PPOCoreConfig(learning_rate=-0.001)  # Must be positive
    with pytest.raises(ValidationError):
        PPOCoreConfig(gamma=1.5)  # Must be <= 1.0

def test_yaml_roundtrip(tmp_path):
    """Test save/load from YAML."""
    config = PPOConfig()
    path = tmp_path / "config.yaml"
    config.to_yaml(str(path))
    loaded = PPOConfig.from_yaml(str(path))
    assert config == loaded
```

## Technical Notes

- Use Pydantic v2 features (BaseModel, Field, field_validator)
- Follow project standards: native Python typing, relative imports from bot2
- Config objects should be immutable after creation for thread safety
- Consider adding `model_config = ConfigDict(frozen=True)` for immutability
- Device "auto" should resolve to CUDA if available, else CPU

## References

- [PPO Paper (Schulman et al.)](https://arxiv.org/abs/1707.06347)
- [Pydantic v2 Documentation](https://docs.pydantic.dev/latest/)
- [Stable-Baselines3 PPO Defaults](https://stable-baselines3.readthedocs.io/en/master/modules/ppo.html)
